CRM-19594 Wrong membership updated

eileenmcnaughton opened the pull request:
* [CRM-19594: Wrong Membership Updated](https://issues.civicrm.org/jira/browse/CRM-19594)

eileenmcnaughton commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -52,6 +52,108 @@ class CRM_Contribute_Form_Contribution_Confirm extends CRM_Contribute_Form_Contr
+   * @param $isEmailReceipt
+   * @return mixed
+   */
+  public static function handlePledge(&$form, $params, $contributionParams, $pledgeID, $contribution, $isEmailReceipt) {
```
This is all just an extraction - moving a chunk of code out of the main function

eileenmcnaughton commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -1445,6 +1460,7 @@ protected function postProcessMembership(
         }
       }
 
+      $membershipParams['skipLineItem'] = 1;
```
ie don't add line items while doing contribution

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -321,6 +321,9 @@ public static function create(&$params, &$ids, $skipRedirect = FALSE, $activityT
       );
     }
 
+    if (empty($params['line_item']) && !empty($params['membership_type_id']) && empty($params['skipLineItem'])) {
```
Moved here from api

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -339,7 +342,12 @@ public static function create(&$params, &$ids, $skipRedirect = FALSE, $activityT
+        $membership->id,
+        $params['line_item'],
+        CRM_Utils_Array::value('contribution', $params),
+        'civicrm_membership'
```
pass in entity

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -1835,48 +1844,20 @@ public static function renewMembership($contactID, $membershipTypeID, $is_test,
         // CRM-15475
         array_search('Cancelled', CRM_Member_PseudoConstant::membershipStatus(NULL, " name = 'Cancelled' ", 'name', FALSE, TRUE)),
       ))) {
-        $membership = new CRM_Member_DAO_Membership();
```
Instead of bypassing membership.create call it with basic params, including the contribution

eileenmcnaughton commented:
@ginkgomzd  I was able to replicate the issue in a test - this fixes the test in the 2 scenarios covered. Really we want more of them..

eileenmcnaughton commented:
@ginkgomzd I'll need you to QA this before I can merge it. I'm worried about what happens to the line items under different combos - I have only ensured memberships are good

eileenmcnaughton commented:
sorry - single item membership pricesets

PalanteJon commented:
Per the request of @eileenmcnaughton, I'm also taking a look at this to see if I can extend test coverage to handle the other cases she mentioned - but this is a lot to wrap my head around.  I'll report in in a few hours if I've made any progress.

eileenmcnaughton commented:
test this please

PalanteJon commented:
I tried extending the test `api_v3_ContributionPageTest::testSubmitMembershipBlockIsSeparatePayment()` to test for this bug as well.  However, it seems that before I even modify the test, it's creating two memberships but only one line item (for the membership, not the donation).

I tested that by adding these 3 lines to the bottom of the test:
```
    CRM_Core_Error::debug_var('contributions', $contributions);
    $lineItems = $this->callAPISuccess('line_item', 'get', array());//, array('contribution_id.contribution_page_id' => $this->_ids['contribution_page']));
    CRM_Core_Error::debug_var('lineItems', $lineItems);
```

If someone confirms that this is what's happening, it seems like it's not really possible to write assertions that pass in this test without fixing that bug first.

KarinG commented:
Just added this to Mattermost - but pasting it bottom-line of that here as well:

I just had a look at public function testSubmitMembershipBlockIsSeparatePayment() { and: 

a) good news is that on a new civibuild dmaster via GUI - doing separate Membership Payment + Contribution produces: 2 entries in the civicrm_contribution and 2 entries in civicrm_line_item; so it's working as advertised. 

However b) when I run this test however - I get: 2 correct entries in the civicrm_contribution and only get a line_item entry for the Membership Amount and it's $1 (and it should be $2). So this test is quite broken - at least on the contribution line_item level. **The actual CiviCRM code is correct though** - it's really just the test that's failing; the way in which it creates the Contributions does not mimic how in the real code base these are (correctly) created.

eileenmcnaughton commented:
@KarinG @PalanteJon thanks for your input. I feel it's important to get this fix right (I have also set the super-observant Tamar onto it).

I have added to the test inline with what @PalanteJon tried and made a change in the code that 'feels right' to make that pass (a change to what I had just added). I'm surprised you couldn't replicate a problem in the UI @KarinG - my sense if the problem would have occured under some circumstance. Any please keep testing / trying to extend the code!

eileenmcnaughton commented:
I am not comfortable that the review on this has been adequate but it either needs to be in or out for the rc and due to the critical nature of the bugs I believe it needs to be in.

I have emailed the money list & both @ginkgomzd  and @davejenx have committed to doing further testing on the rc release so I am merging it

eileenmcnaughton commented:
Since the rc was cut we have had the QA workshop. I tested the following:

Default Contribution Page - with paypal std - paid $1 on quick config. Afterwards checked 1 line item exists for the contribution using firebug + 
```
CRM.api3('LineItem', 'get', {'contribution_id': 95});

Result looks OK & there was exactly one line in the financial transaction pane which showed the correct amount + payment method.


```
97
:
Object
contribution_id
:
"95"
contribution_type_id
:
"1"
entity_id
:
"95"
entity_table
:
"civicrm_contribution"
financial_type_id
:
"1"
id
:
"97"
label
:
"Supporter"
line_total
:
"5.00"
non_deductible_amount
:
"0.00"
participant_count
:
"0"
price_field_id
:
"2"
price_field_value_id
:
"3"
qty
:
"1.00"
unit_price:"5.00"

I got the transactions by doing
```
CRM.api3('FinancialTrxn', 'get', {'options' : {'sort' : 'id DESC', 'limit' : 3}});
```
The 3rd one was unrelated (correct)

The other 2 seemed ok - 

Object
currency:"USD"
fee_amount:"0.00"
id:"94"
is_payment:"1"
net_amount:"5.00"
payment_instrument_id:"1"
payment_processor_id:"5"
status_id
:
"1"
to_financial_account_id
:
"12"
total_amount
:
"5.00"
trxn_date
:
"2016-11-22 01:31:27"
trxn_id
:
"7YF61068V24140155"
__proto__
:
Object
95
:
Object
currency
:
"USD"
fee_amount
:
"0.00"
from_financial_account_id
:
"12"
id
:
"95"
is_payment
:
"1"
net_amount
:
"0.00"
payment_processor_id
:
"5"
status_id
:
"1"
to_financial_account_id
:
"5"
total_amount
:
"0.50"
trxn_date
:
"2016-11-22 01:31:27"
trxn_id
:
"7YF61068V24140155"
__proto__
:
Object

davejenx commented:
### First tested dmaster without PR 9390, before it was committed:
HEAD = 7f57eafa255b6912bb5a882f530e47bdec231e05

**1. Member Signup and Renewal contrib page (no price set) as anon, test payment processor, starting with standard demo db.**

Result: CRM-19594 not replicated, because in standard demo db, highest membership id was 30, highest contribution id was 94 so no id clash.

Created & deleted membership id 95 in MySQL to set auto-increment so that membership ids will overlap with contribution ids.
Created General membership on admin contact in Civi backend, id 96, no contribution.
Then:
**2. Member Signup and Renewal contrib page (no price set) as anon, test payment processor.**

Result: CRM-19594 replicated, as expected:
contribution id 96 & membership id 97 were created;
membership_log showed:
2 new records for existing membership id 96 - WRONG
2 new records for newly created membership id 97 - correct
Also membership_payment had 2 records for contribution id 96:
1 for existing membership id 96 (which previously had no linked contribution) - WRONG
1 for newly created membership id 97 - correct

### Then tested with PR 9390, after it was committed:
- HEAD = 0569a0a5e7047859930d433c5dbe18cb57b9c048

Adjusted dates of previous memberships, before making each contribution, so that I could tell if they were erroneously updated.

**3. Member Signup and Renewal contrib page (no price set) as anon, test payment processor.**

Result: CRM-19594 not replicated: existing memberships not updated.

Created:

1 contact, id 205

1 membership, id 98

2 membership_log:
mysql> select * from civicrm_membership_log where id > 38;
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| id | membership_id | status_id | start_date | end_date   | modified_id | modified_date | membership_type_id | max_related |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| 39 |            98 |         5 | NULL       | NULL       |         205 | 2016-11-22    |                  2 |        NULL |
| 40 |            98 |         1 | 2016-11-22 | 2017-11-21 |         205 | 2016-11-22    |                  2 |        NULL |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
2 rows in set (0.00 sec)
- OK.

1 contribution, id 97

1 membership_payment:
mysql> select * from civicrm_membership_payment where contribution_id = 97;
+----+---------------+-----------------+
| id | membership_id | contribution_id |
+----+---------------+-----------------+
| 34 |            98 |              97 |
+----+---------------+-----------------+
1 row in set (0.00 sec)

1 line_item:
mysql> select * from civicrm_line_item where contribution_id = 97\G
*************************** 1. row ***************************
                   id: 100
         entity_table: civicrm_membership
            entity_id: 98
      contribution_id: 97
       price_field_id: 4
                label: Student
                  qty: 1.00
           unit_price: 50.00
           line_total: 50.00
    participant_count: NULL
 price_field_value_id: 8
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
1 row in set (0.00 sec)
- looks OK.

2 financial transactions:
mysql> select * from civicrm_financial_trxn where id > 97\G
*************************** 1. row ***************************
                       id: 98
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-22 16:54:39
             total_amount: 50.00
               fee_amount: 0.00
               net_amount: 50.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_3_58346a4048fe8
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 2. row ***************************
                       id: 99
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-22 16:54:39
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_3_58346a4048fe8
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
2 rows in set (0.00 sec)

Checking other memberships:

mysql> select id, status_id, start_date, end_date, membership_type_id from civicrm_membership where id > 95;
+----+-----------+------------+------------+--------------------+
| id | status_id | start_date | end_date   | membership_type_id |
+----+-----------+------------+------------+--------------------+
| 96 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
| 97 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 98 |         1 | 2016-11-22 | 2017-11-21 |                  2 |
+----+-----------+------------+------------+--------------------+
3 rows in set (0.00 sec)
- OK.

This all looks OK to me.

I then added an optional additional Other Amount to the Member Signup and Renewal contrib page, not using (explicit) price sets.
**4. Member Signup and Renewal contrib page (no price set) + additional $5 contrib as anon, test payment processor.**

Result: CRM-19594 not replicated: existing memberships not updated.
Query about possible missing line item.

Created:

1 contact, id 206

1 membership, id 99

membership_log:
mysql> select * from civicrm_membership_log where id > 41;
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| id | membership_id | status_id | start_date | end_date   | modified_id | modified_date | membership_type_id | max_related |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| 42 |            99 |         5 | NULL       | NULL       |         206 | 2016-11-22    |                  2 |        NULL |
| 43 |            99 |         1 | 2016-11-22 | 2017-11-21 |         206 | 2016-11-22    |                  2 |        NULL |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
2 rows in set (0.00 sec)

1 contribution, id 98

1 membership_payment:
mysql> select * from civicrm_membership_payment where contribution_id = 98;
+----+---------------+-----------------+
| id | membership_id | contribution_id |
+----+---------------+-----------------+
| 35 |            99 |              98 |
+----+---------------+-----------------+
1 row in set (0.01 sec)

1 line_item:
mysql> select * from civicrm_line_item where contribution_id = 98\G
*************************** 1. row ***************************
                   id: 101
         entity_table: civicrm_membership
            entity_id: 99
      contribution_id: 98
       price_field_id: 4
                label: Student
                  qty: 1.00
           unit_price: 50.00
           line_total: 50.00
    participant_count: NULL
 price_field_value_id: 8
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
1 row in set (0.00 sec)
**- WRONG? Should this be 2 line items, 1 for membership & 1 for $5 additional contrib? There's nothing showing the $5.**

2 financial_trxn:
mysql> select * from civicrm_financial_trxn where id > 99\G
*************************** 1. row ***************************
                       id: 100
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-22 17:20:20
             total_amount: 55.00
               fee_amount: 0.00
               net_amount: 55.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_4_583470446fe93
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 2. row ***************************
                       id: 101
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-22 17:20:20
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_4_583470446fe93
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
2 rows in set (0.00 sec)

Checking other memberships:

mysql> select id, status_id, start_date, end_date, membership_type_id from civicrm_membership where id > 95;
+----+-----------+------------+------------+--------------------+
| id | status_id | start_date | end_date   | membership_type_id |
+----+-----------+------------+------------+--------------------+
| 96 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
| 97 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 98 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 99 |         1 | 2016-11-22 | 2017-11-21 |                  2 |
+----+-----------+------------+------------+--------------------+
4 rows in set (0.00 sec)
- OK.

This mostly looks OK to me except I expected 2 line items to be created: 1 for membership, 1 for additional amount. The one line item created ($50) does not include the $5 additional amount.

I then set up a membership price set with 2 fields: membership + donation.
Set up new contrib page using this.

**5. Test new contrib page (price set, membership + donation) as anon, test payment processor.**

Result: CRM-19594 not replicated: existing memberships not updated.
Query about possible missing line item.

Created:

1 contact, id 207

1 membership, id 100

2 membership_log:
mysql> select * from civicrm_membership_log where id > 43;
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| id | membership_id | status_id | start_date | end_date   | modified_id | modified_date | membership_type_id | max_related |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| 44 |           100 |         5 | NULL       | NULL       |         207 | 2016-11-22    |                  2 |        NULL |
| 45 |           100 |         1 | 2016-11-22 | 2017-11-21 |         207 | 2016-11-22    |                  2 |        NULL |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
2 rows in set (0.00 sec)

1 contribution, id 99

1 membership_payment:
mysql> select * from civicrm_membership_payment where contribution_id = 99;
+----+---------------+-----------------+
| id | membership_id | contribution_id |
+----+---------------+-----------------+
| 36 |           100 |              99 |
+----+---------------+-----------------+
1 row in set (0.00 sec)

1 line_item:
mysql> select * from civicrm_line_item where contribution_id = 99\G
*************************** 1. row ***************************
                   id: 102
         entity_table: civicrm_membership
            entity_id: 100
      contribution_id: 99
       price_field_id: 4
                label: Student
                  qty: 1.00
           unit_price: 50.00
           line_total: 50.00
    participant_count: NULL
 price_field_value_id: 8
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
1 row in set (0.00 sec)
- WRONG? Should this be 2, 1 for membership & 1 for $5 donation? There's nothing showing the $5.

financial_trxn:
mysql> select * from civicrm_financial_trxn where id > 101\G
*************************** 1. row ***************************
                       id: 102
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-22 17:56:32
             total_amount: 55.00
               fee_amount: 0.00
               net_amount: 55.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_5_583478c0ec412
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 2. row ***************************
                       id: 103
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-22 17:56:32
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_5_583478c0ec412
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
2 rows in set (0.00 sec)

As for (4): this mostly looks OK to me except I expected 2 line items to be created: 1 for membership, 1 for donation. The one line item created ($50) does not include the $5 donation.


I then created a new membership org, a new membership type using this & a new price field using that.

**6. Test new contrib page (price set, 2 memberships + donation) as anon, test payment processor.**

Result: CRM-19594 not replicated: existing memberships not updated.
Query about possible missing line item.

Created:
1 contact, id 209

2 memberships, id 101 & 102

4 membership_log:
mysql> select * from civicrm_membership_log where id > 47;
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| id | membership_id | status_id | start_date | end_date   | modified_id | modified_date | membership_type_id | max_related |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| 48 |           101 |         5 | NULL       | NULL       |         209 | 2016-11-22    |                  2 |        NULL |
| 49 |           102 |         5 | NULL       | NULL       |         209 | 2016-11-22    |                  4 |        NULL |
| 50 |           101 |         1 | 2016-11-22 | 2017-11-21 |         209 | 2016-11-22    |                  2 |        NULL |
| 51 |           102 |         1 | 2016-01-01 | 2016-12-31 |         209 | 2016-11-22    |                  4 |        NULL |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
4 rows in set (0.00 sec)

1 contribution, id 100

2 membership_payment:
mysql> select * from civicrm_membership_payment where id > 36;
+----+---------------+-----------------+
| id | membership_id | contribution_id |
+----+---------------+-----------------+
| 37 |           101 |             100 |
| 38 |           102 |             100 |
+----+---------------+-----------------+
2 rows in set (0.01 sec)

2 line_item:
mysql> select * from civicrm_line_item where contribution_id = 100\G
*************************** 1. row ***************************
                   id: 103
         entity_table: civicrm_membership
            entity_id: 101
      contribution_id: 100
       price_field_id: 4
                label: Student
                  qty: 1.00
           unit_price: 50.00
           line_total: 50.00
    participant_count: NULL
 price_field_value_id: 8
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
*************************** 2. row ***************************
                   id: 104
         entity_table: civicrm_membership
            entity_id: 102
      contribution_id: 100
       price_field_id: 13
                label: Test CRM-19594
                  qty: 1.00
           unit_price: 30.00
           line_total: 30.00
    participant_count: NULL
 price_field_value_id: 26
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
2 rows in set (0.01 sec)
- WRONG? Should this be 3, 2 for membership & 1 for $5 donation? There's nothing showing the $5.

financial_trxn:
mysql> select * from civicrm_financial_trxn where id > 103\G
*************************** 1. row ***************************
                       id: 104
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-22 18:32:43
             total_amount: 85.00
               fee_amount: 0.00
               net_amount: 85.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_6_5834813ba41f3
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 2. row ***************************
                       id: 105
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-22 18:32:43
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_6_5834813ba41f3
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
2 rows in set (0.00 sec)

Checking other memberships:
mysql> select id, status_id, start_date, end_date, membership_type_id from civicrm_membership where id > 95;                                                                                                 +-----+-----------+------------+------------+--------------------+
| id  | status_id | start_date | end_date   | membership_type_id |
+-----+-----------+------------+------------+--------------------+
|  96 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
|  97 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
|  98 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
|  99 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 100 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 101 |         1 | 2016-11-22 | 2017-11-21 |                  2 |
| 102 |         1 | 2016-01-01 | 2016-12-31 |                  4 |
+-----+-----------+------------+------------+--------------------+
7 rows in set (0.00 sec)
- OK.

This mostly looks OK to me except I expected 3 line items to be created: 2 for memberships, 1 for donation. The one line item created ($50) does not include the $5 donation.

eileenmcnaughton commented:
Thanks - this is great testing - I'll try to figure out how to replicate your 'wrong' ones in unit tests

eileenmcnaughton commented:
As @davejenx noted there were some gaps. I think I have replicated & fixed enough of them they are hopefully all fixed.

Please TEST. This one matters & the code involved is so toxic there are dead fish floating in it - and dead kittens who ate them. Look at line items, transaction records, totals & create creation of recurrings on any scenario you can think of.

davejenx commented:
@eileenmcnaughton Looks like the latest code changes are on a different PR and a different branch: #9444 and 4.7.14-rc, so I've checked out that branch and am doing some testing.

davejenx commented:
@eileenmcnaughton Ah, I've now spotted that there's also #9449 (open) in addition to #9444 (merged) - I've been testing on the current state of the 4.7.14-rc branch, which doesn't have #9449 because that's not merged. Think it would be worth clarifying what code people should be testing - maybe good to comment on CRM-19594? #9449 isn't currently listed there.

I'll post my results so far with #9444 on that PR, then grab #9449 and start testing that.

eileenmcnaughton commented:
@davejenx it is all merged into the rc now - I wanted to make sure everyone was testing the latest version. Not sure when the tarball will update (within 24 hours but not sure exactly) but the branch should be good now

KarinG commented:
@eileenmcnaughton -I've got 4.7.14-rc on my full iATS test site:

This is what we're transacting:
![membcont01](https://cloud.githubusercontent.com/assets/5340555/20694563/60ec8a64-b5a3-11e6-930b-63916ce4bb29.png)

Note the Type = Member Dues (which is correct) on the Contribution Level (and also in the View Contribution screen) but Type = Donation (which is wrong) [it's the default of the Contribution Page] in the Payment attached to it:
![membcont02](https://cloud.githubusercontent.com/assets/5340555/20694566/654e223e-b5a3-11e6-8512-988333889008.png)

KarinG commented:
**civicrm_contribution looks good:**

| 3581 |          7 |                 1 |                   19 |                     1 | 2016-11-28 21:42:19 |                  NULL |         3.33 |       0.00 |       3.33 | A9535988:1480387340              | c0c497b9d997690fb413fa32cd47e5e2-2 | CAD      | NULL        | 0             | 2016-11-28 21:42:21 | NULL          | Online Contribution: 

| 3582 |          7 |                 2 |                   19 |                     1 | 2016-11-28 21:42:20 |                  NULL |        10.00 |       0.00 |      10.00 | A9535986:1480387339              | 35a0de91f22255b56427a0e79a123304   | CAD      | NULL        | 0             | 2016-11-28 21:42:21 | NULL          | Online Contribution: 

**civicrm_line_item does not:** 3582: was $10 - civicrm_line_item for contribution_id = 3582 has $3.33 - and has it as an add on Donation - financial_type id = 1 (Donation) should be Member Dues. 

MariaDB [civicrm_semperit_com]> select * from civicrm_line_item where contribution_id=3581;
| 3872 | civicrm_contribution |      3581 |            3581 |              1 | Contribution Amount | 1.00 |       3.33 |       3.33 |              NULL |                    1 |                 1 |                  0.00 |       NULL |
+------+----------------------+-----------+-----------------+----------------+---------------------+------+------------+------------+-------------------+----------------------+-------------------+-----------------------+------------+
1 row in set (0.00 sec)

MariaDB [civicrm_semperit_com]> select * from civicrm_line_item where contribution_id=3582;
| 3873 | civicrm_contribution |      3582 |            3582 |             63 | add on Donation | 1.00 |       3.33 |       3.33 |                 0 |                  114 |                 1 |                  0.00 |       NULL |
+------+----------------------+-----------+-----------------+----------------+-----------------+------+------------+------------+-------------------+----------------------+-------------------+-----------------------+------------+
1 row in set (0.00 sec)

KarinG commented:
Let me know if this is helpful and/or if you'd like me to look at other bits.

KarinG commented:
And the receipt is completely wrong:

Receipt #1: $3.33
Receipt #2: $20:

![receipt_wrong](https://cloud.githubusercontent.com/assets/5340555/20694942/d4fd323a-b5a5-11e6-8cea-4f6a4a868e17.png)

KarinG commented:
For sake of completeness: monies in the bank of iATS: total: $13.33:

![iats_bank](https://cloud.githubusercontent.com/assets/5340555/20694997/2cccb79c-b5a6-11e6-9ab0-0547efe46b7c.png)

eileenmcnaughton commented:
@KarinG how are you transacting this? Contribution page configured with membership + separate payment?

eileenmcnaughton commented:
Or price set?

KarinG commented:
![membsep](https://cloud.githubusercontent.com/assets/5340555/20695356/994eb260-b5a8-11e6-9eb9-774482459023.png)

eileenmcnaughton commented:
So, you have configured it as a price set on a contribution page?

KarinG commented:
No in the Configuration it's:

Memberships tab: Membership Section Enabled and
Separate Membership Payment checked
Amounts tab: Contribution Amounts Enabled:
Fixed Contribution Options:
add on Donation $3.33

KarinG commented:
So not using any Price Sets [not in the Memberships tab and not in the Contribution tab of the Contribution Page config]

eileenmcnaughton commented:
Whoever added that separate payment option has a lot to answer for - I swear it has cost us high-five digits over the years!

KarinG commented:
I am successfully using it for a 4.6 project though: College of HCP: Membership fee + Late fee (and I have some code that then automatically decides the correct Late fee based on date/time and hides the other options) - and there it's working on a Contribution and Receipting level [not worried about line_items or financial_items for this project];

KarinG commented:
Not even sure if this PR broke that - or if this is completely different.

eileenmcnaughton commented:
It could well be different - but I'll look anyway

KarinG commented:
The receipt issue for sure is a regression [something that was working in 4.6] - but this may not be the issue here or caused by this PR. Every time I think I understand a few more things about CiviContribute I prove myself wrong...

eileenmcnaughton commented:
@KarinG I have merged a fix into the rc - hopefully it works now....

davejenx commented:
Just checked my previous testing notes and I hadn't re-tested with "Separate Membership Payment" ticked after applying #9449, so it looks like that re-broke it. Have now done that test, with #9449 but without #9464 and confirmed @KarinG's results:

### Re-testing #9449 without #9464

15. Member Signup and Renewal contrib page (no price set) + additional contrib as anon, test payment processor.
"Separate Membership Payment" ticked in contrib page settings.

**Summary**
membership, membership_log, membership_payment OK;
line_items: BROKEN on several counts:
  - line_total: both have $5, should be one $50 and one $5.
  - label: one should be Student but we have Contribution Amount + Additional contribution
  - entity_table: one should have 'civicrm_membership' but both have 'civicrm_contribution'
  - financial_type_id: both are 1 = Donation, one should be 2 = Member Dues

### Testing with latest RC branch code, #9464 merged

17. Member Signup and Renewal contrib page (no price set) + additional contrib as anon, test payment processor.
"Separate Membership Payment" ticked in contrib page settings.

**Summary**
Looks good. :-)

**Gory details**
Created:

1 contact, id 218

1 membership, id 114

2 membership_log:
mysql> select * from civicrm_membership_log where id > 84;
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| id | membership_id | status_id | start_date | end_date   | modified_id | modified_date | membership_type_id | max_related |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
| 85 |           114 |         5 | NULL       | NULL       |         218 | 2016-11-29    |                  2 |        NULL |
| 86 |           114 |         1 | 2016-11-29 | 2017-11-28 |         218 | 2016-11-29    |                  2 |        NULL |
+----+---------------+-----------+------------+------------+-------------+---------------+--------------------+-------------+
2 rows in set (0.00 sec)
- OK

2 contributions, id 112 & 113

1 membership_payment:
mysql> select * from civicrm_membership_payment where contribution_id in (112, 113);
+----+---------------+-----------------+
| id | membership_id | contribution_id |
+----+---------------+-----------------+
| 48 |           114 |             113 |
+----+---------------+-----------------+
1 row in set (0.00 sec)
- OK

2 line_items:
mysql> select * from civicrm_line_item where contribution_id in (112, 113) order by contribution_id, id\G                                                                                                                                    
*************************** 1. row ***************************
                   id: 123
         entity_table: civicrm_contribution
            entity_id: 112
      contribution_id: 112
       price_field_id: 1
                label: Contribution Amount
                  qty: 1.00
           unit_price: 5.00
           line_total: 5.00
    participant_count: NULL
 price_field_value_id: 1
    financial_type_id: 1
non_deductible_amount: 0.00
           tax_amount: NULL
*************************** 2. row ***************************
                   id: 124
         entity_table: civicrm_membership
            entity_id: 114
      contribution_id: 113
       price_field_id: 5
                label: Student
                  qty: 1.00
           unit_price: 50.00
           line_total: 50.00
    participant_count: 0
 price_field_value_id: 11
    financial_type_id: 2
non_deductible_amount: 0.00
           tax_amount: NULL
2 rows in set (0.00 sec)
- OK.

4 financial_trxn:
mysql> select * from civicrm_financial_trxn where trxn_date >= '2016-11-29 16:40'\G
*************************** 1. row ***************************
                       id: 128
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-29 16:40:31
             total_amount: 50.00
               fee_amount: 0.00
               net_amount: 50.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_10_583da16f89f4b
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 2. row ***************************
                       id: 129
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-29 16:40:31
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_10_583da16f89f4b
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
*************************** 3. row ***************************
                       id: 130
from_financial_account_id: NULL
  to_financial_account_id: 12
                trxn_date: 2016-11-29 16:40:31
             total_amount: 5.00
               fee_amount: 0.00
               net_amount: 5.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_10_583da16fbf1c7
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: 1
             check_number: NULL
*************************** 4. row ***************************
                       id: 131
from_financial_account_id: 12
  to_financial_account_id: 5
                trxn_date: 2016-11-29 16:40:31
             total_amount: 1.50
               fee_amount: 0.00
               net_amount: 0.00
                 currency: USD
               is_payment: 1
                  trxn_id: live_10_583da16fbf1c7
         trxn_result_code: NULL
                status_id: 1
     payment_processor_id: 1
    payment_instrument_id: NULL
             check_number: NULL
4 rows in set (0.00 sec)

Checking other memberships:
mysql> select id, contact_id, status_id, start_date, end_date, membership_type_id from civicrm_membership where id >= 95;
+-----+------------+-----------+------------+------------+--------------------+
| id  | contact_id | status_id | start_date | end_date   | membership_type_id |
+-----+------------+-----------+------------+------------+--------------------+
|  96 |        202 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
|  97 |        204 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
|  98 |        205 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
|  99 |        206 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 100 |        207 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 101 |        209 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 102 |        209 |         4 | 2015-04-01 | 2016-03-31 |                  4 |
| 103 |        210 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
| 104 |        211 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
| 105 |        212 |         4 | 2015-04-01 | 2016-03-31 |                  1 |
| 106 |        213 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 107 |        214 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 108 |        214 |         4 | 2015-04-01 | 2016-03-31 |                  4 |
| 109 |        215 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 110 |        216 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 111 |        217 |         4 | 2015-04-01 | 2016-03-31 |                  2 |
| 112 |        202 |         4 | 2015-04-01 | 2016-03-31 |                  4 |
| 113 |        200 |         4 | 2015-04-01 | 2016-03-31 |                  4 |
| 114 |        218 |         1 | 2016-11-29 | 2017-11-28 |                  2 |
+-----+------------+-----------+------------+------------+--------------------+
19 rows in set (0.05 sec)
- OK

KarinG commented:
Ok - happy dance on Type = Member Dues now [for the Membership]:

![typememberdues](https://cloud.githubusercontent.com/assets/5340555/20739046/dff9ae46-b674-11e6-8bbf-98ae7daef5fc.png)

KarinG commented:
Email receipt still shows double/double:

![doubleamount](https://cloud.githubusercontent.com/assets/5340555/20739121/481e7e02-b675-11e6-8e5b-c702cf22c843.png)

KarinG commented:
civicrm_contribution: fine - ids: 3584 $3.33 and 3585 $10
civicrm_line_item: happy dance here too:

![lineitemshappy](https://cloud.githubusercontent.com/assets/5340555/20739439/d3b3bb2e-b677-11e6-8d26-27838b71a72e.png)

davejenx commented:
Hmm, I didn't receive email receipts so didn't see that issue. Something up with mail on localhost. :-(

litespeedmarc commented:
For anyone still on or about to go live on 4.7.13 (and unable/unwilling to upgrade at the very last minute), is there a workaround or something that can be done to avoid this?   It sounds like it happens when membership IDs & contribution IDs collide?  Perhaps bump up the contribution ID sequence as a workaround?

davejenx commented:
@litespeedmarc yes, the updating of wrong memberships happens when membership IDs & contribution IDs collide. Bumping up the contribution ID sequence would I think avoid this, as a workaround. I think line items could still get created with incorrect entity_ids even so, with entity_table = 'civicrm_membership' but with entity_id = the contribution id instead of the membership id. But if there's no membership with that id, at least it avoids also updating the wrong membership.

eileenmcnaughton commented:
I think you might still get line item inaccuracies. I would encourage you to test the rc.

eileenmcnaughton commented:
@KarinG  @davejenx I'm starting to think the wrong email details on separate payment issue might be present in 4.7.13 rather than being a 4.7.14 regression - can either of you confirm that?

If so I'm inclined to NOT attempt to fix this for 4.7.14 but give it more time in 4.7.15 - the code seems particularly gnarly and if it's actually a NEW can of worms I might not open it just yet
